subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/result/VersionSelectionReasons.java

```diff
@@ -17,83 +17,127 @@
-            return CONFLICT_RESOLUTION_BY_RULE;
-        } else if (reason == REQUESTED) {
-            return CONFLICT_RESOLUTION;
-        } else if (reason == FORCED) {
-            return CONFLICT_RESOLUTION;
-        } else if (reason == ROOT) {
-            return reason;
-        }
-        throw new IllegalArgumentException("Cannot create conflict resolution selection reason for input: " + reason);
+    public static final ComponentSelectionDescriptorInternal REQUESTED = new DefaultComponentSelectionDescriptor(ComponentSelectionCause.REQUESTED);
+    public static final ComponentSelectionDescriptorInternal ROOT = new DefaultComponentSelectionDescriptor(ComponentSelectionCause.ROOT);
+    public static final ComponentSelectionDescriptorInternal FORCED = new DefaultComponentSelectionDescriptor(ComponentSelectionCause.FORCED);
+    public static final ComponentSelectionDescriptorInternal CONFLICT_RESOLUTION = new DefaultComponentSelectionDescriptor(ComponentSelectionCause.CONFLICT_RESOLUTION);
+    public static final ComponentSelectionDescriptorInternal SELECTED_BY_RULE = new DefaultComponentSelectionDescriptor(ComponentSelectionCause.SELECTED_BY_RULE);
+    public static final ComponentSelectionDescriptorInternal COMPOSITE_BUILD = new DefaultComponentSelectionDescriptor(ComponentSelectionCause.COMPOSITE_BUILD);
+
+    public static ComponentSelectionReasonInternal requested() {
+        return new DefaultComponentSelectionReason(REQUESTED);
+    }
+
+
+    public static ComponentSelectionReason root() {
+        return new DefaultComponentSelectionReason(ROOT);
+    }
+
+    public static ComponentSelectionReasonInternal of(List<ComponentSelectionDescriptor> descriptions) {
+        return new DefaultComponentSelectionReason(descriptions);
     }
 
     private static class DefaultComponentSelectionReason implements ComponentSelectionReasonInternal {
 
-        private final boolean forced;
-        private final boolean conflictResolution;
-        private final boolean selectedByRule;
-        private final boolean expected;
-        private final boolean compositeParticipant;
-        private final String description;
+        // Use a set to de-duplicate same descriptions
+        private final Set<ComponentSelectionDescriptorInternal> descriptions;
 
-        private DefaultComponentSelectionReason(boolean forced, boolean conflictResolution, boolean selectedByRule, boolean expected, boolean compositeBuild, String description) {
-            this.forced = forced;
-            this.conflictResolution = conflictResolution;
-            this.selectedByRule = selectedByRule;
-            this.expected = expected;
-            this.compositeParticipant = compositeBuild;
-            assert description != null;
-            this.description = description;
+        private DefaultComponentSelectionReason(ComponentSelectionDescriptor description) {
+            descriptions = Sets.newLinkedHashSet();
```

Comment:
`LinkedHashSet` is particularly expensive: I wouldn't be surprised if an `ImmutableSet` wasn't more efficient (for read speed and memory consumption). Construction might be a little slower, but for the small sets we are working with I'd be surprised.